### PR TITLE
soc: nrf53: add pretick workaround for anomaly 165

### PR DESCRIPTION
**Pull request by ankuns**
This PR adds workaround for the anomaly 165 for nRF53 in part related to wake up on RTC timers, commonly called "pretick".

When the `arch_cpu_idle` or `arch_cpu_atomic_idle` is entered, the RTC1 and RTC0 are examined for their CCs and OVERFLOW events that are enabled and able to generate interrupt. The tick at which the closest event is scheduled is searched. The dedicated CC channel on RTC1 called pretick CC is set up to trigger one tick before.
The pretick CC triggers WDT and IPC with DPPI according to D5.WA3 of Nordic Semiconductor's internal documentation.
The pretick CC itself does not generate interrupt.
The possible difference between RTC0 and RTC1 is handled, they are not required to be started in sync.

The examination of RTCs and setting up the pretick takes time (around 10us). If it was done with interrupts disabled directly in `z_arm_on_enter_cpu_idle` hook, this could have detrimental impact on time-critical activities such as BLE.

Because of that `z_arm_on_enter_cpu_idle_prepare` hook is introduced. It is called before interrupts are disabled. It's role is to examine RTCs and set up pretick CC. To ensure that between start of `z_arm_on_enter_cpu_idle_prepare` and processing of it's outcome in `z_arm_on_enter_cpu_idle` no exception happened that could potentially change RTCs the LDREX/STREX operations are used and the fact that on exception exit local monitor is cleared.

The initial implementation of pretick CC setup and DPPI network according to D5.WA3 is taken form the cherry-picked commit https://github.com/nordic-krch/fw-nrfconnect-zephyr/commit/a0016385c41c9a6528f8ea98b5ec3cd30fd9495c .
It has been then improved by:
- requiring just one RTC1 CC for pretick purpose
- not imposing any requirements (including reservation of any additional CCs) on RTC0 users other than calling `rtc_pretick_rtc0_isr_hook` from RTC0 ISR.
- allowing user channels for nrf_rtc_timer
- allowing nRF 802.15.4 Radio Driver

The presence of the workaround is controlled by Kconfig option `SOC_NRF53_RTC_PRETICK`. 
It is turned on by default for nRF5340, a user can turn it off.
The workaround uses watchdog on network core directly, the workaround is turned off when the user uses requests to use watchdog driver.

### CODE REVIEW

**Review by ahasztag — APPROVED** [review]

**Comment by microbuilder** [comment]
Arm-specific code looks good, but please clear up the compliance issues.

**Comment by ankuns** [comment]
@microbuilder code style issues fixed

**Review by microbuilder — DISMISSED** [review]
The Arm part of this looks fine to me, though unable to test the nRF side on HW. Presumably has been vetted in house, though, at Nordic.

**Review by martintv — APPROVED** [review]

**Comment by microbuilder** [comment]
@carlescufi Can someone with approval rights from Nordic have a look before the freeze?

**Review by anangl — CHANGES_REQUESTED** [review]

File: arch/arm/core/cortex_m/cpu_idle.S [hunk 1]
```diff
@@ -92,6 +92,12 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
 #endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
 #endif /* CONFIG_TRACING */
 
+#if defined(CONFIG_ARM_ON_ENTER_CPU_IDLE_PREPARE_HOOK)
+	push	{r0, lr}
+	bl	z_arm_on_enter_cpu_idle_prepare
+	pop	{r0, lr}
```
[the remark this answers is not in the document]
Comment (microbuilder) [reply arch/arm/core/cortex_m/cpu_idle.S:98]: @anangl Thanks for catching this 🤦‍♂️

[the remark this answers is not in the document]
Comment (ankuns) [reply arch/arm/core/cortex_m/cpu_idle.S]: @anangl fixed in 9199877b2a797f31f87ff5ab9af9ed5af7a5c0ec, please have a look.

File: soc/arm/nordic_nrf/nrf53/soc.c [hunk 2]
```diff
@@ -267,32 +463,21 @@ static int rtc_pretick_cpunet_init(void)
...
+	nrf_rtc_event_clear(NRF_RTC1, NRF_RTC_CHANNEL_EVENT_ADDR(RTC1_PRETICK_CC_CHAN));
 
 	return 0;
 }
 #endif /* CONFIG_SOC_NRF5340_CPUNET */
 
-static int rtc_pretick_init(const struct device *unused)
+static int rtc_pretick_init(void)
 {
-	ARG_UNUSED(unused);
```
[the remark this answers is not in the document]
Comment (ankuns) [reply soc/arm/nordic_nrf/nrf53/soc.c]: Corrections moved to the commit you mentioned, please have a look into f6dce6dccc36231dc801a9e04df8be6f37519b53

File: soc/arm/nordic_nrf/nrf53/soc.c [hunk 3]
```diff
@@ -123,25 +134,359 @@ static bool nrf53_anomaly_160_check(void)
...
+
+	nrfx_gppi_task_endpoint_setup(ch, task_ipc);
+	nrfx_gppi_event_endpoint_setup(ch, evt_ipc);
+	nrfx_gppi_channels_enable(BIT(ch));
+
+	return 0;
+}
+#else /* CONFIG_SOC_NRF5340_CPUNET */
+
+void rtc_pretick_rtc_isr_hook(void)
```
[the remark this answers is not in the document]
Comment (ankuns) [reply soc/arm/nordic_nrf/nrf53/soc.c]: done, in the commit that introduces it, please have a look into f6dce6dccc36231dc801a9e04df8be6f37519b53

File: soc/arm/nordic_nrf/nrf53/soc.c [hunk 4]
```diff
@@ -123,25 +134,359 @@ static bool nrf53_anomaly_160_check(void)
...
+{
+	bool result = false;
+
+	/* Let's preload registers to speed-up. The function is not thread-safe anyway. */
+	uint32_t reg_intenset = rtc->INTENSET;
+
+	/* If corresponding EVTEN bit is zero, then interrupt won't be generated as the event
+	 * itself wont be generated.
+	 */
+	reg_intenset &= rtc->EVTEN;
```
[the remark this answers is not in the document]
Comment (ankuns) [reply soc/arm/nordic_nrf/nrf53/soc.c]: You are right, great thanks for spotting it! I missed this behavior, fixed in e7ea373a5d8dc41e0d946994185de93e2a8181b3

File: soc/arm/nordic_nrf/nrf53/soc.c [hunk 5]
```diff
@@ -123,25 +134,359 @@ static bool nrf53_anomaly_160_check(void)
...
+	do {
+		*counter_a = nrf_rtc_counter_get(rtc_a);
+		barrier_dmem_fence_full();
+		*counter_b = nrf_rtc_counter_get(rtc_b);
+		barrier_dmem_fence_full();
+	} while (*counter_a != nrf_rtc_counter_get(rtc_a));
+}
+#endif
+
+#if defined(CONFIG_SOC_NRF53_RTC_PRETICK) && defined(CONFIG_SOC_NRF5340_CPUNET)
```
[the remark this answers is not in the document]
Comment (ankuns) [reply soc/arm/nordic_nrf/nrf53/soc.c]: done

File: soc/arm/nordic_nrf/nrf53/soc.c [hunk 6]
```diff
@@ -123,25 +134,359 @@ static bool nrf53_anomaly_160_check(void)
...
+			if (nrf_rtc_event_check(NRF_RTC1,
+					NRF_RTC_CHANNEL_EVENT_ADDR(RTC1_PRETICK_CC_CHAN))) {
+				NRF_WDT->TASKS_START = 1;
+			}
+		}
+	}
+
+	cpu_idle_prepare_allows_sleep = ok_to_sleep;
+}
+#endif
```
[the remark this answers is not in the document]
Comment (ankuns) [reply soc/arm/nordic_nrf/nrf53/soc.c]: done

File: soc/arm/nordic_nrf/nrf53/soc.c [hunk 7]
```diff
@@ -123,25 +134,359 @@ static bool nrf53_anomaly_160_check(void)
...
 
 	if (ok_to_sleep) {
-		suppress_message = false;
-	} else if (!suppress_message) {
-		LOG_DBG("Anomaly 160 trigger conditions detected.");
-		suppress_message = true;
+		ok_to_sleep = nrf53_anomaly_160_check();
+
+	#if (LOG_LEVEL >= LOG_LEVEL_DBG)
```
[the remark this answers is not in the document]
Comment (ankuns) [reply soc/arm/nordic_nrf/nrf53/soc.c]: done

**Comment by ankuns** [comment]
@anangl, I fixed all nits you pointed, please have a look.

**Review by anangl — APPROVED** [review]

**Review by microbuilder — APPROVED** [review]

**Pull request merged**